### ISSUE TITLE
Increase Quote Precision from 4 places to 8 places

### ISSUE THREAD

**Issue by metal450** [buchen/portfolio#1796]
**Is your feature request related to a problem? Please describe.**
Currently, PP is capable of quotes with 4 digits precision.  However, there are quite a few cryptocurrencies whose value is less than 0.0001 (example: [KIN](https://www.livecoinwatch.com/price/Kin-KIN) is 0.000023).  Since these quotes aren't representable with just 4 places precision, the quote gets rounded to zero, & thus no matter how much you own it can only show 0 value.  For instance, if you own 12 million Kin, it has value of $276 - but since the quote is < 4 digits, it's not possible to represent, and PP always shows it as worth $0, and thus 100% loss for every such position.

**Describe the solution you'd like**
Please increase quote precision from 4 decimal places to 8, to make PP more widely compatible with cryptocurrencies :)

As originally discussed a bit in this issue: https://github.com/buchen/portfolio/issues/1459

**Comment by metal450** [comment]
Question: respectfully, if I were to contribute some financial amount, might there be a way to get this & #1692 bumped up for sooner consideration?

I absolutely love this software, but do struggle greatly with trying to compensate for this issue by adding & adjusting manual 'placeholder' values to compensate for the cryptos that can't be represented, and/or whose quotes need to be repeatedly input manually.  These 2 issues alone would make the daily workflow immeasurably more streamlined :)

**Comment by Drkmstr310** [comment]
I have same problem - many tokens with small values result in either strongly inaccurate totals, or completely zero (i.e. RAVEN, BTT, etc).  Is it quite difficult to add some digits? :)

**Comment by buchen** [comment]
That is a fair feature request. Let me quickly explain why it is not so easy.

10 years ago, when I started PP, I decided to store all values in cents. The rationale was that with integers (or more specifically longs) I get the right precision while simplifying code and saving memory compared to using BigDecimal objects. By now, PP knows how to handle fractions of a cent in the same way (after all quotes can have 4 decimal places). Unfortunately there are already lots of cases where I am close to or reaching the memory limit of longs. So adding more decimal places to the quote is not so easy because a) changing everything to BigDecimal is a lot of work throughout the code base, b) the file size increases quite a lot if I just add more fractions (historical quotes are already a big chunk of the data).

Nonetheless, I have started to dig through the code to get a better understanding of the impact on the change. As always, I cannot promise anything right now, but I am aware of this deficiency.

**Comment by metal450** [comment]
Thanks for the reply.

I understand - & I imagined it might be something like that.  Again, I absolutely love PP & it has become such an indispensable part of my daily workflow - just this one issue has been such a sticking point that I actually did dedicate a few hours to trying to dig into the code myself (especially as crypto allocation has grown & net worth becomes increasingly inaccurate).  I totally see how it isn't a quick fix - thus just even getting reply to this thread was honestly a little exciting (what a nerd I am!) :P

As for the issue of file size, I wonder if this could be solved with a simple option?  Obviously the code itself would still need to be updated to support BigDecimal, and I get that that's the bulk of the work, but as far as file size - I could envision each security having a checkbox "store high-precision quotes."  By default, quotes would still just serialize the current number of digits.  Only if the user deliberately enables the setting for a given security would it serialize the extra digits & thus would filesize be affected.
